### PR TITLE
Enabling the cancellation of non-actor tasks in a worker's queue

### PR DESCRIPTION
**Pull request by goliaro**
## Why are these changes needed?

These changes are needed to fix issue [#12080](https://github.com/ray-project/ray/issues/12080). 

With the new code, `CoreWorker::HandleCancelTask` is able to cancel non-actor tasks that are waiting (in a worker's queue) for their turn to execute. The changes do not affect how actor tasks are cancelled.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -2167,19 +2167,26 @@ void CoreWorker::HandleCancelTask(const rpc::CancelTaskRequest &request,
...
+  if (kill_success && !request.force_kill()) {
     RAY_LOG(INFO) << "Interrupting a running task " << main_thread_task_id_;
-    success = options_.kill_main();
+    kill_success = options_.kill_main();
+  } else if (!kill_success) {
+    
+    // If the task is not currently running, check if it is in the worker's queue of normal tasks, and remove it if found. 
+    if (direct_task_receiver_->CancelQueuedNormalTask(task_id)) {
+      success=true;
+    }
```
Comment (ijrsvt) [inline src/ray/core_worker/core_worker.cc:2182]: Can you just assign?  `success = direct_task_receiver_->CancelQueuedNormalTask(task_id)`?

File: src/ray/core_worker/core_worker.cc [hunk 2]
```diff
@@ -2167,19 +2167,26 @@ void CoreWorker::HandleCancelTask(const rpc::CancelTaskRequest &request,
                                   rpc::SendReplyCallback send_reply_callback) {
   absl::MutexLock lock(&mutex_);
   TaskID task_id = TaskID::FromBinary(request.intended_task_id());
-  bool success = main_thread_task_id_ == task_id;
+  bool kill_success = main_thread_task_id_ == task_id;
```
Comment (ijrsvt) [inline src/ray/core_worker/core_worker.cc:2170]: Could you rename this to `requested_task_running` (i.e. `requested_task_running= main_thread_task_id_ == task_id;`

File: src/ray/core_worker/core_worker.cc [hunk 3]
```diff
@@ -2167,19 +2167,26 @@ void CoreWorker::HandleCancelTask(const rpc::CancelTaskRequest &request,
...
+  bool success = kill_success;
 
   // Try non-force kill
-  if (success && !request.force_kill()) {
+  if (kill_success && !request.force_kill()) {
     RAY_LOG(INFO) << "Interrupting a running task " << main_thread_task_id_;
-    success = options_.kill_main();
+    kill_success = options_.kill_main();
```
Comment (ijrsvt) [inline src/ray/core_worker/core_worker.cc:2176]: This should just be `success`

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 4]
```diff
@@ -383,6 +389,11 @@ class ActorSchedulingQueue : public SchedulingQueue {
     ScheduleRequests();
   }
 
+  // We don't cancel actor tasks by removing them from the actor queues, so we don't do anything here.
```
Comment (ijrsvt) [inline src/ray/core_worker/transport/direct_actor_transport.h:392]: nit: We don't cancel any actor tasks, so maybe:
```
// We don't cancel actor tasks, so we don't do anything here. 
```

**Review by ijrsvt — CHANGES_REQUESTED** [review]
Looks great! Thanks so much for doing this really quickly. I left a few comments and was also thinking it would be good to add a few CPP tests for this!

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 5]
```diff
@@ -383,6 +389,9 @@ class ActorSchedulingQueue : public SchedulingQueue {
     ScheduleRequests();
   }
 
+  // We don't cancel actor tasks, so we don't do anything here.
+  bool CancelTaskIfFound(TaskID task_id) { return false; }
```
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_actor_transport.h:393]:
```suggestion
  bool CancelTaskIfFound(TaskID task_id) { RAY_CHECK(false) << "Cannot cancel actor tasks"; }
```

**Review by stephanie-wang — APPROVED** [review]
Thanks, this looks great! Just left a suggestion to make cancelling an actor task fatal.

**Comment by stephanie-wang** [comment]
+1 on adding a unit test!

**Comment by gabrieleoliaro** [comment]
@ijrsvt @stephanie-wang I just added two unit tests!

**Review by ijrsvt — APPROVED** [review]
Looks good! Thanks for adding some unit tests :)

**Comment by stephanie-wang** [comment]
Looks like there is a related CI test timeout:
```
==================== Test output for //python/ray/tests:test_cancel:
============================= test session starts ==============================
platform linux -- Python 3.6.12, pytest-5.4.3, py-1.9.0, pluggy-0.13.1 -- /opt/miniconda/bin/python3
cachedir: .pytest_cache
rootdir: /home/travis/.cache/bazel/_bazel_travis/b88c129a127452fc94033a29d9f90e20/execroot/com_github_ray_project_ray/bazel-out/k8-opt/bin/python/ray/tests/test_cancel.runfiles/com_github_ray_project_ray
plugins: asyncio-0.14.0, rerunfailures-9.1.1, sugar-0.9.4, timeout-1.4.2
collecting ... collected 15 items

::test_cancel_chain[True] PASSED                                         [  6%]
::test_cancel_chain[False] PASSED                                        [ 13%]
::test_cancel_multiple_dependents[True] PASSED                           [ 20%]
::test_cancel_multiple_dependents[False] PASSED                          [ 26%]
::test_single_cpu_cancel[True] PASSED                                    [ 33%]
::test_single_cpu_cancel[False] PASSED                                   [ 40%]
::test_comprehensive[True] PASSED                                        [ 46%]
::test_comprehensive[False] PASSED                                       [ 53%]
::test_stress[True] -- Test timed out at 2020-11-28 07:07:09 UTC --
```

Also, could you fix the lint (`./ci/travis/format.sh`)?

**Comment by gabrieleoliaro** [comment]
I'm working on the timeouts! When it comes to linting, I'm confused as to why the test fails, because the linter script seems to show that the code is already lint-ed:

`(myenv) dhcp-10-250-248-49:ray Gabriele$ ./ci/travis/format.sh 
WARNING: Ray uses mypy 0.782, You currently are using 0.790. This might generate different results.
WARNING: clang-format is not installed!
WARNING: Ray uses flake8 with flake8_quotes. Might error without it. Install with: pip install flake8-quotes
From github.com:ray-project/ray
 * branch                master     -> FETCH_HEAD
(myenv) dhcp-10-250-248-49:ray Gabriele$ git status
On branch cancel_queued_tasks
Your branch is up to date with 'origin/cancel_queued_tasks'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	python/ray/dashboard/

nothing added to commit but untracked files present (use "git add" to track)
`

**Comment by gabrieleoliaro** [comment]
@ijrsvt @stephanie-wang Does anyone have any advice on how to debug the `test_cancel` timeout? In the Travis build which fails, it looks like the timeout is caused by `test_stress[True]` within `/python/ray/tests/test_cancel.py`. The` test_cancel.py` pytest shows some weird behavior on my local machine. In particular, `test_stress[True]` always passes (both on the PR branch as well as on the master branch), whereas `test_fast[False]` always gets stuck (it never runs to completion, with low cpu/memory utilization). I'm particularly confused by the fact that `test_fast[False]` doesn't finish even when it is run on master:
test_cancel

I can try merging the master branch in again, with `git merge master`, but it looks like the latest commits on ray-project:master also have some Travis builds that fail, so idk if it would help.

The good news is that at least the `CI / windows-msvc (pull_request)` tests are passing now (see the "linting" commit)!

Thanks!

**Comment by stephanie-wang** [comment]
It's possible `test_fast` was already flaky. I did see one test failure on flakey-tests.ray.io but it seems to fail infrequently. I think `test_stress` is an actual regression since it's timing out on all of the builds for this run. Sometimes it's difficult to reproduce Travis logs, but you could try running the pytest command with `taskset` to pin to one core. Something like this:
`taskset 1 pytest python/ray/tests/test_cancel.py::test_stress[True]`

@ijrsvt, any pointers on these test cases?

**Comment by gabrieleoliaro** [comment]
Yeah, looking at the code again, test_stress seems likely to be caused by a deadlock. On the other hand, I'm still super confused about test_fast. Let me try running the test again with the `taskset` option

**Comment by ijrsvt** [comment]
Hmm, that is weird! I am able to reproduce the problem with `test_fast[False]` locally. It appears to hang in the final getting of results.

From some poking around it appears that the failure is in the first `100` results, which means that somehow a task is getting stuck?
I also noticed that while getting here (adding print statements to `test_fast`) that certain print statements made the test more or less flakey.

**Comment by stephanie-wang** [comment]
@gabrieleoliaro @ijrsvt looks like `test_cancel` is still timing out. Any ideas why?

**Comment by gabrieleoliaro** [comment]
> @gabrieleoliaro @ijrsvt looks like `test_cancel` is still timing out. Any ideas why?

It looks like there is some race condition involving the mutex in the `NormalSchedulingQueue` class. In fact, when I commented it out, all tests in `test_cancel` passed quickly. Let me see if the race condition involves all the methods in NormalSchedulingQueue or only some

**Comment by stephanie-wang** [comment]
> It looks like there is some race condition involving the mutex in the `NormalSchedulingQueue` class. In fact, when I commented it out, all tests in `test_cancel` passed quickly. Let me see if the race condition involves all the methods in NormalSchedulingQueue or only some

Ah I see, one tip is if you can reproduce the race condition locally, you can use `gdb` to attach to a hanging process and use `backtrace` to see where it's hanging.

**Comment by gabrieleoliaro** [comment]
> > It looks like there is some race condition involving the mutex in the `NormalSchedulingQueue` class. In fact, when I commented it out, all tests in `test_cancel` passed quickly. Let me see if the race condition involves all the methods in NormalSchedulingQueue or only some
> 
> Ah I see, one tip is if you can reproduce the race condition locally, you can use `gdb` to attach to a hanging process and use `backtrace` to see where it's hanging.

let me try that!

**Comment by gabrieleoliaro** [comment]
@stephanie-wang @ijrsvt I think I found the problem. While holding the lock, the function NormalSchedulingQueue::ScheduleRequests in direct_actor_transport.h loops through a queue of InboundRequests and calls Accept() on each of them, before popping them from the queue. However, sometimes execution does not come back from this line here: https://github.com/gabrieleoliaro/ray/blob/ff05231dbb3599cb88197ad2c559845f68553391/src/ray/core_worker/transport/direct_actor_transport.h#L569 . So I think I just need to cal head.Accept() without holding the lock.

**Comment by stephanie-wang** [comment]
Ah interesting, I'm not sure why calling Accept() would deadlock though. I guess it must be some interaction with the cancellation code. Were you able to figure out where exactly Accept() was hanging? Calling Accept() outside of the lock sounds good, though.

**Comment by gabrieleoliaro** [comment]
> Ah interesting, I'm not sure why calling Accept() would deadlock though. I guess it must be some interaction with the cancellation code. Were you able to figure out where exactly Accept() was hanging? Calling Accept() outside of the lock sounds good, though.

I just tested the code after changing the loop in NormalSchedulingQueue::ScheduleRequests() to call Accept() after releasing the lock. While the logs now suggest that the execution no longer gets stuck within Accept(), that threads no longer hang while grabbing a lock or hang/exit/freeze while holding a lock, the tests still get stuck. So I don't think the change fixed the problem (although it might have removed some additional bugs). Will keep investigating! Let me know if you have any guess of what could be the cause. I still think it's some synchronization problem, because everything worked smoothly when I commented out the mutex & locking instructions

**Comment by gabrieleoliaro** [comment]
The interesting thing is -- the deadlock seems to be present in the master branch as much as it is in this PR's branch, so it might actually be an issue that is still lingering from the [Queueing non-actor tasks at the workers](https://github.com/ray-project/ray/pull/11051) PR

**Comment by gabrieleoliaro** [comment]
@ijrsvt @stephanie-wang Hi everyone, and happy New Year! After experimenting a bit more over the last few days, I am now almost certain that the timeout in the `python/ray/tests/test_cancel.py::test_stress[False]` is independent from the PR. In fact, after noticing that the `test_cancel `pytest was getting stuck locally even with the master branch, I created a new branch (see [here](https://github.com/gabrieleoliaro/ray/tree/revert_PR_11051)) where I reverted the changes from this PR as well as my previous PR #11051). The test still hangs at `python/ray/tests/test_cancel.py::test_stress[False]` (see screenshot below), so the cause cannot be in the new code. 

test_cancel_hangs

After running the `python/ray/tests/test_cancel.py::test_stress[False]` on the new branch (where I had reverted the changes of the two PRs), I also took a look at the core-driver logs and I noticed that they also contained a lot of the errors:

`W1209 00:03:50.420928 69159 422989248 reference_count.cc:195] Tried to decrease ref count for nonexistent object ID: 711f7fe0b401aa0cffffffff0100000001000000`

Hence, these warnings, too, should be independent from the new changes

**Comment by stephanie-wang** [comment]
Okay sounds good! I'll merge it for now and keep an eye on the flakey tests. We'll revert this PR if this does increase the flakiness.

**Pull request merged**